Remove extra contexts while deopting

JanJecmen opened the pull request.

JanJecmen committed:
cleaning up

o- commented:
Here are some stats for your PR:

* WARNING: the longest CI job test_release_2 took 3.05h
* Bounce_nonames_simple regressed by 0.94
* spectralnorm_naive regressed by 0.83
* spectralnorm_alt_3 regressed by 0.93
* fasta regressed by 0.95
* matrixFor regressed by 0.93
* Overall benchmarks regressed by 0.99

Please find your performance results at https://rir-benchmarks.prl.fit.cvut.cz/diff?job_ids[]=2573576996&job_ids[]=2547932342&selection=all

JanJecmen committed:
more cleanup

JanJecmen committed:
Remove extra contexts while deopting

Because of an optimization where we replace the first checkpoint in an inlinee by the checkpoint at the callsite it is possible to take the outer deopt exit from within the inlinee. In case the inlinee needs its own context we need to pop it, otherwise the context offsets in MkEnv instructions in the deopt block would modify the wrong frame.
```
a = checkpoint -> [deopt with (mkenv ... context 1)]
...
c = push_context ....
assume ..., a -> we have to pop c
```
The other option would be to make the context field in MkEnvs a PIR variable and then set it to the correct value based on where we come from, however, this seems easier.

o- commented:
Here are some stats for your PR:

* WARNING: the longest CI job test_features_1 took 2.95h
* spectralnorm_alt_3 regressed by 0.94
* matrixFor regressed by 0.93
* Overall benchmarks regressed by 1.0

Please find your performance results at https://rir-benchmarks.prl.fit.cvut.cz/diff?job_ids[]=2583039279&job_ids[]=2578982488&selection=all
